draft-ietf-quic-ack-frequency.md

```diff
@@ -323,10 +323,17 @@ acknowledgements.
 
 ## Expediting Congestion Signals {#congestion}
 
-As specified in Section 13.2.1 of {{QUIC-TRANSPORT}}, an endpoint SHOULD
-immediately acknowledge packets marked with the ECN Congestion Experienced (CE)
-codepoint in the IP header. Doing so reduces the peer's response time to
-congestion events.
+And endpoing SHOULD send an immediate acknowledgement when a packet marked
+with the ECN Congestion Experienced (CE) codepoint in the IP header is
+received after the previous packet was not marked CE.
+
+An endpoint SHOULD also send an immediate acknowledgement when at least
+max(2, Ack-Eliciting Threshold) packets marked with the ECN Congestion
+Experienced (CE) codepoint in the IP header are received. Doing so reduces the
```

Comment:
How about
> An endpoint SHOULD reduce its Ack-Eliciting Threshold to at most 2 when acknowledging a contiguous series of CE marked packets.

Reply:
I think the original attempt to emulate Accurate ECN feedback in TCP might have been misguided. The AccECN machinery was necessary in TCP because TCP does _not_ have ACK frequency control. Here in QUIC we are adding ACK frequency control. So the only addition needed for ECN should indeed be to immediately trigger an ACK on a transition from non-CE to CE marking. Then, QUIC should just continue ACKing every Ack-Eliciting Threshold packets. No need for the max(2, ...). I think (I'm not working on QUIC, so I don't keep the whole of the QUIC design in my working memory).

That sorts out the lower bound on ACK frequency, but it leaves one problem at the upper bound (if I understand QUIC correctly). [RFC9000 §13.2.1.  Sending ACK Frames](https://datatracker.ietf.org/doc/html/rfc9000#section-13.2.1) and [§13.2.4. Limiting Ranges by Tracking ACK Frames](https://datatracker.ietf.org/doc/html/rfc9000#section-13.2.4) suggest that a host that is only sending non-ack-eliciting packets could occasionally include an ack-eliciting packet (e.g. a ping) to catch up on the feedback and allow the other end to clear out the state its been collecting.

An example where this could be a problem is when data transfer switches back and forth from one QUIC peer to the other. Imagine an asymmetric data transfer from A to B ends. If the network CE-marked many or all the ACKs on the reverse path (B-A), I believe A will not ACK any of the last round of ACKs it receives. Then if B becomes the sender of a volley of data, it will unknowingly send data into a path that was already congested just with ACKs.

With ECN, I'm not sure it's the best design to rely on the sender of the non-ack-eliciting packets (B in the example) to send an occasional ack-eliciting packet. If the receiver of the non-ack-eliciting packets (A) is seeing a large amount of ECN marks, it ought to be able to send occasional ACKs of non-ack-eliciting packets (to B), whether or not B asks for them.

In AccECN, we allowed TCP to ack a sequence of CE-marked ACKs, as long as the ACK frequency was no less than 1 per 3 CE-marked packets. Even with 100% CE-marking, this damps any potential ACK ping-pong but still ensures the other end receive feedback of congestion events.

Reply:
Thanks, I'll update this to make it so CE marked packets are treated like ack-eliciting packets from an ACK perspective, which I think solves the upper bound problem you're outlining?